### PR TITLE
xds interop: update interop client to support new tests

### PR DESCRIPTION
**Pull request by menghanl**
- Add flags `--rpc` and `--metadata`
 - Update stats service to report `rpc_by_type`
 - Change client and server to read/send hostname in metadata, instead of RPC response

### CODE REVIEW

File: interop/grpc_testing/test.proto [hunk 1]
```diff
@@ -214,6 +214,13 @@ message LoadBalancerStatsRequest {
 }
 
 message LoadBalancerStatsResponse {
+  message RPCsByPeer {
+    // The number of completed RPCs for each peer.
+    map<string, int32> rpcs_by_peer = 1;
+  }
+  // The number of completed RPCs for each type (UnaryCall or EmptyCall).
+  map<string, RPCsByPeer> rpcs_by_type = 3;
```
Comment (ericgribkoff) [inline interop/grpc_testing/test.proto:222]: Nit: These are out of order now (3 is before 1 and 2). Move this block to after `num_failures`?

Comment (menghanl) [reply interop/grpc_testing/test.proto]: Done. Also fixed in https://github.com/grpc/grpc/pull/23436

File: interop/grpc_testing/test.proto [hunk 2]
```diff
@@ -214,10 +214,17 @@ message LoadBalancerStatsRequest {
 }
 
 message LoadBalancerStatsResponse {
+  message RPCsByPeer {
```
Comment (ericgribkoff) [inline interop/grpc_testing/test.proto:217]: I believe the proto style here is for this to be camel-cased: RpcsByPeer.

Comment (menghanl) [reply interop/grpc_testing/test.proto]: Done

File: interop/xds/client/client.go [hunk 3]
```diff
@@ -39,18 +42,43 @@ type statsWatcherKey struct {
...
 	rpcsByPeer    map[string]int32
+	rpcsByType    map[rpcType]map[string]int32
 	numFailures   int32
 	remainingRpcs int32
-	c             chan *testpb.SimpleResponse
+	chanHosts     chan *rpcInfo
+}
+
+func (watcher *statsWatcher) buildResp() *testpb.LoadBalancerStatsResponse {
+	rpcsByType := make(map[string]*testpb.LoadBalancerStatsResponse_RpcsByPeer)
```
Comment (dfawley) [inline interop/xds/client/client.go:61]: Optional, but usually a good idea: `make(map[]..., len(watcher.rpcsByType))`

...or, would it be better to store `rpcsByType` as `map[type-string]RpcsByPeer` so it doesn't need to be re-formed at all?

File: interop/xds/client/client.go [hunk 4]
```diff
@@ -95,25 +124,90 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+	if len(rpcs) == 0 {
+		return []rpcType{"UnaryCall"}
+	}
+
+	for _, r := range rpcs {
+		switch r {
+		case "UnaryCall":
+			ret = append(ret, unaryCall)
+		case "EmptyCall":
+			ret = append(ret, emptyCall)
```
Comment (dfawley) [inline interop/xds/client/client.go:169]: `case "UnaryCall", "EmptyCall": ret = append(ret, rpcType(r))`?

File: interop/xds/client/client.go [hunk 5]
```diff
@@ -95,25 +124,90 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+
+const (
+	unaryCall rpcType = "UnaryCall"
+	emptyCall rpcType = "EmptyCall"
+)
+
+func parseRPCTypes(rpcStr string) (ret []rpcType) {
+	rpcs := strings.Split(rpcStr, ",")
+	if len(rpcs) == 0 {
+		return []rpcType{"UnaryCall"}
```
Comment (dfawley) [inline interop/xds/client/client.go:161]: `return []rpcType{unaryCall}`

File: interop/xds/client/client.go [hunk 6]
```diff
@@ -95,25 +124,90 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+type rpcType string
+
+const (
+	unaryCall rpcType = "UnaryCall"
+	emptyCall rpcType = "EmptyCall"
+)
+
+func parseRPCTypes(rpcStr string) (ret []rpcType) {
+	rpcs := strings.Split(rpcStr, ",")
+	if len(rpcs) == 0 {
```
Comment (dfawley) [inline interop/xds/client/client.go:160]: Optional: `if len(rpcStr) == 0` and move above the `Split`.

File: interop/xds/client/client.go [hunk 7]
```diff
@@ -95,25 +124,90 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+	rpcMetadataSplit := strings.Split(rpcMetadataStr, ",")
+	rpcsToMD := make(map[string][]string)
+	for _, rm := range rpcMetadataSplit {
+		rmSplit := strings.Split(rm, ":")
+		if len(rmSplit)%2 != 1 {
+			log.Fatalf("invalid metadata config %v, want EmptyCall:key1:value1", rm)
+		}
+		rpcsToMD[rmSplit[0]] = append(rpcsToMD[rmSplit[0]], rmSplit[1:]...)
+	}
+	var ret []*rpcConfig
```
Comment (dfawley) [inline interop/xds/client/client.go:195]: `ret := make([]*rpcConfig, len(rpcs))`

File: interop/xds/client/client.go [hunk 8]
```diff
@@ -135,16 +229,52 @@ func main() {
...
 }
 
-func sendRPCs(clients []testpb.TestServiceClient, ticker *time.Ticker) {
+func makeOneRPC(c testpb.TestServiceClient, cfg *rpcConfig) (*peer.Peer, *rpcInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), *rpcTimeout)
+	defer cancel()
+	var (
+		p      peer.Peer
+		header metadata.MD
+	)
```
Comment (dfawley) [inline interop/xds/client/client.go:241]: if you are going to use a var block, maybe move it down to where `err` is declared and include that too?

File: interop/xds/client/client.go (hunk 3 above)
Comment (menghanl) [reply interop/xds/client/client.go:61]: Done. Added length.
And removed `rpcType` all together.

File: interop/xds/client/client.go (hunk 6 above)
Comment (menghanl) [reply interop/xds/client/client.go:160]: Done

File: interop/xds/client/client.go (hunk 5 above)
Comment (menghanl) [reply interop/xds/client/client.go:161]: Done

File: interop/xds/client/client.go (hunk 4 above)
Comment (menghanl) [reply interop/xds/client/client.go:169]: Done

File: interop/xds/client/client.go (hunk 7 above)
Comment (menghanl) [reply interop/xds/client/client.go:195]: Done

File: interop/xds/client/client.go (hunk 8 above)
Comment (menghanl) [reply interop/xds/client/client.go:241]: Done

File: interop/xds/client/client.go [hunk 9]
```diff
@@ -95,25 +124,86 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+}
+
+const (
+	unaryCall string = "UnaryCall"
+	emptyCall string = "EmptyCall"
+)
+
+func parseRPCTypes(rpcStr string) (ret []string) {
+	if len(rpcStr) == 0 {
+		return []string{"UnaryCall"}
```
Comment (dfawley) [inline interop/xds/client/client.go:158]: `unaryCall`

File: interop/xds/client/client.go [hunk 10]
```diff
@@ -95,25 +124,86 @@ func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalanc
...
+	}
+
+	rpcs := strings.Split(rpcStr, ",")
+	for _, r := range rpcs {
+		switch r {
+		case "UnaryCall", "EmptyCall":
```
Comment (dfawley) [inline interop/xds/client/client.go:164]: `case unaryCall, emptyCall`

File: interop/xds/client/client.go (hunk 9 above)
Comment (menghanl) [reply interop/xds/client/client.go]: Done

File: interop/xds/client/client.go (hunk 10 above)
Comment (menghanl) [reply interop/xds/client/client.go]: Done

**Pull request merged**